src/common/client.py

```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""EtcdClient utility class to connect to etcd server and execute commands with etcdctl."""
+
+import json
+import logging
+import subprocess
+
+logger = logging.getLogger(__name__)
+
+
+class EtcdClient:
+    """Handle etcd client connections and run etcdctl commands."""
+
+    def __init__(
+        self,
+        client_url: str,
+    ):
+        self.client_url = client_url
+
+    def get_endpoint_status(self) -> dict:
+        """Run the `endpoint status` command and return the result as dict."""
+        endpoint_status = {}
+        if result := self._run_etcdctl(
+            command="endpoint",
+            subcommand="status",
+            endpoints=self.client_url,
+            output_format="json",
+        ):
+            endpoint_status = json.loads(result)[0]
+
+        return endpoint_status
+
+    def _run_etcdctl(
+        self,
+        command: str,
+        subcommand: str | None,
+        endpoints: str,
+        output_format: str | None = "simple",
+    ) -> str | None:
+        """Execute `etcdctl` command via subprocess.
+
+        The list of arguments will be extended once authentication/encryption is implemented.
+        This method aims to provide a very clear interface for executing `etcdctl` and minimize
+        the margin of error on cluster operations.
+
+        Args:
+            command: command to execute with etcdctl, e.g. `elect`, `member` or `endpoint`
+            subcommand: subcommand to add to the previous command, e.g. `add` or `status`
+            endpoints: str-formatted list of endpoints to run the command against
+            output_format: set the output format (fields, json, protobuf, simple, table)
+            ...
+
+        Returns:
+            The output of the subprocess-command as a string. In case of error, this will
+            return `None`. It will not raise an error in order to leave error handling up
+            to the caller. Depending on what command is executed, the ways of handling errors
+            might differ.
+        """
+        try:
+            result = subprocess.run(
+                args=[
+                    "etcdctl",
+                    command,
+                    subcommand,
+                    f"--endpoints={endpoints}",
+                    f"-w={output_format}",
+                ],
+                check=True,
+                capture_output=True,
+                text=True,
+            ).stdout
```

Comment:
```suggestion
            ).stdout.strip()
```

Reply:
This is not required here, as we use `.stdout`. It would cause an error instead:

```
subprocess.run(["charmed-etcd.etcdctl", "member", "list", "--endpoints=10.54.237.224:2379", "-w=simple"]).stdout.strip()
3c64c92fd7480501, started, etcd50, http://10.54.237.252:2380, http://10.54.237.252:2379, false
75a3fc47cc32bceb, started, etcd48, http://10.54.237.224:2380, http://10.54.237.224:2379, false
75b9c521876ed7e5, started, etcd49, http://10.54.237.117:2380, , false
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'strip'
```

Reply:
@reneradoi the above command fails as you're not capturing the output of the command - it is simply printing it.
However for the setup you have with `capture_output=True, text=True` will succeed. 
```
>>> subprocess.run(["charmed-etcd.etcdctl", "member", "list", "--endpoints=localhost:2379", "-w=simple"], check=True, capture_output=True, text=True).stdout
'8e9e05c52164694d, started, default, http://localhost:2380, http://localhost:2379, false\n'
>>> subprocess.run(["charmed-etcd.etcdctl", "member", "list", "--endpoints=localhost:2379", "-w=simple"], check=True, capture_output=True, text=True).stdout.strip()
'8e9e05c52164694d, started, default, http://localhost:2380, http://localhost:2379, false'
```

Reply:
Sorry, missed that. Thanks for the hint!